File: src/compiler/request.rs
#![allow(clippy::module_name_repetitions)]

use crate::compiler::{CompletionEntry, Diagnostic, Location, ModuleIdentifier, Position};
use std::collections::BTreeMap;

#[derive(Debug, serde::Serialize)]
#[serde(tag = "type", content = "request", rename_all = "snake_case")]
pub enum Request {
	Check(CheckRequest),
	Rename(RenameRequest),
	Diagnostics(DiagnosticsRequest),
	Definition(DefintionRequest),
	Hover(HoverRequest),
	References(ReferencesRequest),
	Completion(CompletionRequest),
	Transpile(TranspileRequest),
}

#[derive(Debug, serde::Deserialize)]
#[serde(tag = "type", content = "response", rename_all = "snake_case")]
pub enum Response {
	Check(CheckResponse),
	Rename(RenameLocationsResponse),
	Diagnostics(DiagnosticsResponse),
	Hover(HoverResponse),
	References(ReferencesResponse),
	Definition(DefinitionResponse),
	Completion(CompletionResponse),
	Transpile(TranspileResponse),
}

#[derive(Debug, serde::Serialize)]
#[serde(rename_all = "camelCase")]
pub struct CheckRequest {
	pub module_identifiers: Vec<ModuleIdentifier>,
}

#[derive(Debug, serde::Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct CheckResponse {
	pub diagnostics: BTreeMap<ModuleIdentifier, Vec<Diagnostic>>,
}

#[derive(Debug, serde::Serialize)]
#[serde(rename_all = "camelCase")]
pub struct RenameRequest {
	pub module_identifier: ModuleIdentifier,
	pub position: Position,
}

#[derive(Debug, serde::Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct RenameLocationsResponse {
	pub locations: Option<Vec<Location>>,
}

#[derive(Debug, serde::Serialize)]
#[serde(rename_all = "camelCase")]
pub struct DiagnosticsRequest {}

#[derive(Debug, serde::Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct DiagnosticsResponse {
	pub diagnostics: BTreeMap<ModuleIdentifier, Vec<Diagnostic>>,
}

#[derive(Debug, serde::Serialize)]
#[serde(rename_all = "camelCase")]
pub struct HoverRequest {
	pub module_identifier: ModuleIdentifier,
	pub position: Position,
}

#[derive(Debug, serde::Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct HoverResponse {
	pub text: Option<String>,
}

#[derive(Debug, serde::Serialize)]
#[serde(rename_all = "camelCase")]
pub struct ReferencesRequest {
	pub module_identifier: ModuleIdentifier,
	pub position: Position,
}

#[derive(Debug, serde::Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct ReferencesResponse {
	pub locations: Option<Vec<Location>>,
}

#[derive(Debug, serde::Serialize)]
#[serde(rename_all = "camelCase")]
pub struct DefintionRequest {
	pub module_identifier: ModuleIdentifier,
	pub position: Position,
}

#[derive(Debug, serde::Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct DefinitionResponse {
	pub locations: Option<Vec<Location>>,
}

#[derive(Debug, serde::Serialize)]
#[serde(rename_all = "camelCase")]
pub struct CompletionRequest {
	pub module_identifier: ModuleIdentifier,
	pub position: Position,
}

#[derive(Debug, serde::Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct CompletionResponse {
	pub entries: Option<Vec<CompletionEntry>>,
}

#[derive(Debug, serde::Serialize)]
#[serde(rename_all = "camelCase")]
pub struct TranspileRequest {
	pub text: String,
}

#[derive(Debug, serde::Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct TranspileResponse {
	pub output_text: String,
	pub source_map_text: String,
}

File: src/compiler/request.ts
import { Location, CompletionEntry, Diagnostic, Position } from "./types";

export type Request =
	| { type: "check"; request: CheckRequest }
	| { type: "completion"; request: CompletionRequest }
	| { type: "rename"; request: RenameRequest }
	| { type: "format"; request: FormatRequest }
	| { type: "diagnostics"; request: DiangosticsRequest }
	| { type: "hover"; request: HoverRequest }
	| { type: "references"; request: ReferencesRequest }
	| { type: "definition"; request: DefinitionRequest }
	| { type: "transpile"; request: TranspileRequest };

export type Response =
	| { type: "check"; response: CheckResponse }
	| { type: "completion"; response: CompletionResponse }
	| { type: "rename"; response: RenameResponse }
	| { type: "format"; response: FormatResponse }
	| { type: "diagnostics"; response: DiagnosticsResponse }
	| { type: "hover"; response: HoverResponse }
	| { type: "references"; response: ReferencesResponse }
	| { type: "definition"; response: DefinitionResponse }
	| { type: "transpile"; response: TranspileResponse };

export type CheckRequest = { moduleIdentifiers: Array<string> };

export type CheckResponse = {
	diagnostics: { [key: string]: Array<Diagnostic> };
};

export type CompletionRequest = {
	moduleIdentifier: string;
	position: Position;
};

export type CompletionResponse = {
	entries?: Array<CompletionEntry>;
};

export type RenameRequest = {
	moduleIdentifier: string;
	position: Position;
};

export type RenameResponse = {
	locations: Array<Location> | null | undefined;
};

export type FormatRequest = {
	text: string;
};

export type FormatResponse = {
	text: string;
};

export type DiangosticsRequest = {};

export type DiagnosticsResponse = {
	diagnostics: { [key: string]: Array<Diagnostic> };
};

export type HoverRequest = {
	moduleIdentifier: string;
	position: Position;
};

export type HoverResponse = {
	text?: string;
};

export type ReferencesRequest = {
	moduleIdentifier: string;
	position: Position;
};

export type ReferencesResponse = {
	locations: Array<Location> | null;
};

export type DefinitionRequest = {
	moduleIdentifier: string;
	position: Position;
};

export type DefinitionResponse = {
	locations: Array<Location> | null;
};

export type TranspileRequest = {
	text: string;
};

export type TranspileResponse = {
	outputText: string;
	sourceMapText: string;
};
